scripts/install-build-tools.sh

```diff
@@ -57,30 +92,85 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
-  $SUDO add-apt-repository "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main"
+  $SUDO apt update -y
   $SUDO apt install -y clang-format-14 clang-tidy-14
-  $SUDO ln -s -f $(which clang-format-14) /usr/local/bin/clang-format
-  $SUDO ln -s -f $(which clang-tidy-14) /usr/local/bin/clang-tidy
+  $SUDO ln -sf $(which clang-format-14) /usr/local/bin/clang-format
+  $SUDO ln -sf $(which clang-tidy-14) /usr/local/bin/clang-tidy
 
-  # Python 3.10+ not installed so get the latest version
-  if ! check_python_version; then
-    $SUDO add-apt-repository ppa:deadsnakes/ppa
-    $SUDO apt update
+  # Install Python and necessary packages
+  $SUDO add-apt-repository -y ppa:deadsnakes/ppa
+  $SUDO apt update -y
 
-    latest_python_version=$(apt-cache search python3. | grep -o 'python3\.[0-9]*' | sort -V | tail -n 1)
-    $SUDO apt install -y $latest_python_version
+  # If default python3 version is already installed and supported, then just install pylint and matplotlib
+  # not going to look through all the user's python versions to see if any are in supported versions
+  if (echo "${PYTHON3_DEFAULT}" | grep -E 'Python 3.1[0-2]'); then
+    FULL_PY="python${PYTHON3_DEFAULT_VERS}"
+    # Check if the desired Python version is available in the apt repository
+    if apt list -a ${FULL_PY} &> /dev/null; then
+      # install packages for the default python3 version which is valid
+      $SUDO apt install -y python${PYTHON3_DEFAULT_VERS}-distutils
+      $SUDO ${FULL_PY} -m pip install pylint matplotlib
+    fi
+    echo "Successfully installed dependencies for default Python3 version: ${PYTHON3_DEFAULT}"
+  # User could have valid python versions installed, but could be using complex env so won't check
+  # Install one of the compatible python versions
+  else
+    PY_INSTALLED=''
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+      # Check if the desired Python version is available in the apt repository
+      FULL_PY="python${PY_VERS}"
+      # Check if the desired Python version is available in the apt repository
+      if apt list -a ${FULL_PY} &> /dev/null; then
+        echo "Installing ${FULL_PY} and dependencies"
+        $SUDO apt install -y python${PY_VERS} python${PY_VERS}-distutils
+        # for dependency installation on specific python version
+        wget https://bootstrap.pypa.io/get-pip.py
+        $SUDO ${FULL_PY} get-pip.py && rm -f get-pip.py
+        $SUDO ${FULL_PY} -m pip install pylint matplotlib
+        PY_INSTALLED="python${PY_VERS}"
+        echo "Successfully installed dependencies for ${FULL_PY}"
+        break
+      fi
+    done
+
+    # post-install checks
+    if [[ "$PY_INSTALLED" == '' ]]; then
+      echo "Python3 install with apt failed, attempted on these versions: ${PYTHON_VERSIONS[@]}"
+      exit 1
+    # if automated flow, we will update the default python3 version, otherwise let user do manually
+    elif [[ $# -eq 1 && $1 == "GITHUB_ACTIONS" ]]; then
```

Comment:
Does the CI fail without this? This is a lot of special-casing.

Reply:
Yes it does

Reply:
Is it wildly more difficult to detect the presence of a python version we expect and then call it directly instead of depending on `python` (or `python3`) being the right version?

Is it reasonable to extend our use of `prefix` to include symlinks to versions of things we like so we can use `prefix/bin/python` instead of `python` and have symlinked it to a sensible version instead?

Special-casing things like this is a maintenance nightmare.

Reply:
I like that idea - will implement this. Will simplify the flow

Reply:
We no longer care about default - the script checks if a valid python install exists. This searched for in the brew path for darwin (mac) and the /usr/bin/ path for linux